### PR TITLE
[SYCL] Improve mutation of literal address space for variadic printf

### PR DESCRIPTION
**Pull request by AGindinson**
The initial implementation of address space mutation assumed that in
presence of `experimental::printf` wrappers at the IR level, each
wrapper would use its dedicated signature of the `__spirv_ocl_printf`
builtin. However, this is not the case when the variadic signature of
the builtin is employed - in that scenario, the wrapper does not
become variadic, so `__spirv_ocl_printf` may have multiple wrapper
instances using it.

Adjust the algorithm for removing generic AS functions accordingly and
add LIT coverage for the non-inline variadic case.

In addition, minimize the IR for the non-variadic O0 test.

Signed-off-by: Artem Gindinson <artem.gindinson@intel.com>

### CODE REVIEW

**Comment by AGindinson** [comment]
There's clearly an issue with code-ownership rules - @intel/dpcpp-tools-reviewers should've been assigned. I'll rebase the PR once this is figured out.

File: llvm/test/SYCLLowerIR/printf_addrspace/generic_as_variadic_no_opt.ll [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+$_ZN2cl4sycl3ext6oneapi12experimental6printfIcJiEEEiPKT_DpT0_ = comdat any
+
+; CHECK-DAG: @.str._AS2 = internal addrspace(2) constant [15 x i8] c"String No. %f\0A\00", align 1
+@.str = private unnamed_addr addrspace(1) constant [15 x i8] c"String No. %f\0A\00", align 1
+; CHECK-DAG: @.str.1._AS2 = internal addrspace(2) constant [15 x i8] c"String No. %i\0A\00", align 1
+@.str.1 = private unnamed_addr addrspace(1) constant [15 x i8] c"String No. %i\0A\00", align 1
+
+; Function Attrs: convergent noinline norecurse optnone
+define linkonce_odr dso_local spir_func void @_ZN2cl4sycl8accessorINS0_6detail14AssertHappenedELi1ELNS0_6access4modeE1025ELNS4_6targetE2014ELNS4_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEC2Ev(%"class.cl::sycl::accessor" addrspace(4)* align 8 dereferenceable_or_null(32) %this) unnamed_addr #1 comdat align 2 {
```
Comment (AlexeySachkov) [inline llvm/test/SYCLLowerIR/printf_addrspace/generic_as_variadic_no_opt.ll:66]: Let's drop assert handling stuff from the test to make it cleaner. That piece of LLVM IR doesn't affect your test anyway.

Comment (AGindinson) [reply llvm/test/SYCLLowerIR/printf_addrspace/generic_as_variadic_no_opt.ll:66]: OK, I'll see if there's a further way to minimize the IR. Here, I've essentially copied the contents of `printf_addrspace/generic_as_no_opt.ll` and made a couple adjustments in the IR/description, and I was under the impression we've already minimized that for the original patch. That said, maybe I can reduce both of these.

Comment (AGindinson) [reply llvm/test/SYCLLowerIR/printf_addrspace/generic_as_variadic_no_opt.ll]: Done within https://github.com/intel/llvm/pull/5286/commits/be2af3667af7a3ef8db2c1d31954cc20b2ee12b1 & https://github.com/intel/llvm/pull/5286/commits/801bbc0184e2935d543002b5ae2d606607267654.

**Review by AlexeySachkov — APPROVED** [review]

**Pull request merged**